### PR TITLE
Concurrent graph copy

### PR DESCRIPTION
**Pull request by cpuguy83**
This is a carry of #38034
Closes #38034

I squashed commits for an easier rebase.

---

- What I did
I made copies concurrent for VFS, and Overlay1.

- How I did it
I started with adding a benchmark to the tests to give me a baseline. On my test system that gave me about ~1081927574 ns/op. I slowly improved the performance throughout the commits, and increased performance to about 552743671 ns/op in the final tests. I parallizes

- How to verify it
There are attached benchmarks, and tests

- Description for the changelog
Make VFS Faster.

### CODE REVIEW

File: daemon/graphdriver/overlay/overlay.go [hunk 1]
```diff
@@ -12,6 +12,8 @@ import (
 	"strconv"
 	"strings"
 
+	"github.com/docker/docker/errdefs"
+
```
Comment (thaJeztah) [inline daemon/graphdriver/overlay/overlay.go]: nit: should be grouped with the other imports below

File: daemon/graphdriver/vfs/driver.go [hunk 2]
```diff
@@ -85,19 +87,27 @@ func (d *Driver) parseOptions(options []string) error {
...
+				return err
 			}
 			if err = d.setQuotaOpt(uint64(size)); err != nil {
-				return errdefs.InvalidParameter(errors.Wrap(err, "failed to set option size for vfs"))
+				return errors.Wrap(err, "failed to set option size for vfs")
+			}
+		case "copy.concurrency":
+			d.copyConcurrency, err = strconv.Atoi(val)
+			if err != nil {
+				return errors.Wrapf(err, "cannot parse copy.concurrency")
```
Comment (thaJeztah) [inline daemon/graphdriver/vfs/driver.go]:
```suggestion
				return errors.Wrap(err, "cannot parse copy.concurrency")
```

File: daemon/graphdriver/overlay/overlay.go [hunk 3]
```diff
@@ -166,27 +163,39 @@ func Init(home string, options []string, uidMaps, gidMaps []idtools.IDMap) (grap
...
 		if err != nil {
-			return nil, err
+			return err
 		}
 		key = strings.ToLower(key)
 		switch key {
+		case "copy.concurrency":
+			d.copyConcurrency, err = strconv.Atoi(val)
+			if err != nil {
+				return errors.Wrapf(err, "cannot parse copy.concurrency")
```
Comment (thaJeztah) [inline daemon/graphdriver/overlay/overlay.go]:
```suggestion
				return errors.Wrap(err, "cannot parse copy.concurrency")
```

**Review by thaJeztah — CHANGES_REQUESTED** [review]
didn't do a full review, but spotted some minor issues when glancing over

**Comment by thaJeztah** [comment]
@dmcgowan @tonistiigi @kolyshkin PTAL 🤗

File: daemon/graphdriver/overlay/overlay.go (hunk 1 above)
Comment (cpuguy83) [reply daemon/graphdriver/overlay/overlay.go]: Fixed

File: daemon/graphdriver/vfs/driver.go (hunk 2 above)
Comment (cpuguy83) [reply daemon/graphdriver/vfs/driver.go]: Fixed

File: daemon/graphdriver/overlay/overlay.go (hunk 3 above)
Comment (cpuguy83) [reply daemon/graphdriver/overlay/overlay.go]: Fixed

**Pull request closed without merging**